packages/rendermime/src/attachmentmodel.ts

```diff
@@ -0,0 +1,252 @@
+  }
+
+  /**
+   * The data associated with the model.
+   */
+  get data(): ReadonlyJSONObject {
+    return this._rawData;
+  }
+
+  /**
+   * The metadata associated with the model.
+   */
+  get metadata(): ReadonlyJSONObject {
+    return undefined;
+  }
+
+  /**
+   * Set the data associated with the model.
+   *
+   * #### Notes
+   * Depending on the implementation of the mime model,
+   * this call may or may not have deferred effects,
+   */
+  setData(options: IRenderMime.IMimeModel.ISetDataOptions): void {
+    if (options.data) {
+      this._updateObservable(this._data, options.data);
+      this._rawData = options.data;
+    }
+    this._changed.emit(void 0);
+  }
+
+  /**
+   * Serialize the model to JSON.
+   */
+  toJSON(): nbformat.IMimeBundle {
+    let attachment: JSONValue = {};
+    for (let key in this._raw) {
+      attachment[key] = Private.extract(this._raw, key);
+    }
+    return attachment as nbformat.IMimeBundle;
+  }
+
+  // All attachments are untrusted
+  readonly trusted = false;
+
+  /**
+   * Update an observable JSON object using a readonly JSON object.
+   */
+  private _updateObservable(observable: IObservableJSON, data: ReadonlyJSONObject) {
+    let oldKeys = observable.keys();
+    let newKeys = Object.keys(data);
+
+    // Handle removed keys.
+    for (let key of oldKeys) {
+      if (newKeys.indexOf(key) === -1) {
+        observable.delete(key);
+      }
+    }
+
+    // Handle changed data.
+    for (let key of newKeys) {
+      let oldValue = observable.get(key);
+      let newValue = data[key];
+      if (oldValue !== newValue) {
+        observable.set(key, newValue as JSONValue);
+      }
+    }
+  }
+
+  private _changed = new Signal<this, void>(this);
+  private _raw: JSONObject = {};
+  private _rawData: ReadonlyJSONObject;
+  private _data: IObservableJSON;
+}
+
+
+/**
+ * The namespace for AttachmentModel statics.
+ */
+export
+namespace AttachmentModel {
+  /**
+   * Get the data for an attachment.
+   *
+   * @params bundle - A kernel attachment MIME bundle.
+   *
+   * @returns - The data for the payload.
+   */
+  export
+  function getData(bundle: nbformat.IMimeBundle): JSONObject {
+    return Private.getData(bundle);
+  }
+}
+
+
+/**
+ * The namespace for module private data.
+ */
+namespace Private {
+  /**
+   * Get the data from a notebook attachment.
+   */
+  export
+  function getData(bundle: nbformat.IMimeBundle): JSONObject {
+    return convertBundle(bundle);
+  }
+
+  /**
+   * Get the bundle options given attachment model options.
+   */
+  export
+  function getBundleOptions(options: IAttachmentModel.IOptions): MimeModel.IOptions {
+    let data = getData(options.value);
+    return { data };
+  }
+
+  /**
+   * Extract a value from a JSONObject.
+   */
+  export
+  function extract(value: JSONObject, key: string): JSONValue {
+    let item = value[key];
+    if (JSONExt.isPrimitive(item)) {
+      return item;
+    }
+    return JSON.parse(JSON.stringify(item));
```

Comment:
Since you already have `JSONExt`, you can use `JSONExt.deepCopy(item)` here.

Reply:
Also copied directly from `outputmodel`.